Feature/bjs2 40449 Запрос рекомендации

MikhailShap opened the pull request.

MikhailShap committed:
Created classes:
- RecommendationRequest
- RecommendationRequestDto
- RecommendationRequestMapper
- RecommendationRequestMapperTest
- RecommendationRequestService
- RecommendationRequestServiceValidator
- RequestFrequencyException
- SkillNotFoundException
- SkillRequest

Modifications:
- Added annotation @Accessors(chain = true) in RecommendationRequest

Created methods:
- requestRecommendation in RecommendationRequestController
- toDTO, toEntity, and default mapSkillRequestToSkillId in RecommendationRequestMapper
- toDTO, toEntity, and setUp in RecommendationRequestMapperTest
- create in RecommendationRequestService
- validateExistsRequesterAndReceiverInDatabase, validateSixMonthRequestLimit, and validateExistsSkillsInDatabase in RecommendationRequestServiceValidator

Created RecommendationRequestDto with necessary fields

MikhailShap committed:
Removed custom exceptions:RequestFrequencyException,SkillNotFoundException.

Added custom exception:DataValidationException.
Modified RecommendationRequestService:
changed the addition of SkillRequest to the RecommendationRequest, along with the set specified via addSkillRequest

MikhailShap committed:
Modification:
added method getRecommendationRequests in RecommendationRequestController;
added method getRequests in RecommendationRequestService

Created RequestFilterDto with necessary fields

MikhailShap committed:
Modification:
added validate field status in Class RequestFilterDto;
created method allToDTO in RecommendationRequestMapper;
created interface RequestFilter and method apply;
created StatusFilter implementing RequestFilter;

completed the implementation of the logic in the method getRequests in RecommendationRequestService

MikhailShap committed:
Modification:
added method getRecommendationRequest in RecommendationRequestController;
added method getRequest in RecommendationRequestService;

MikhailShap committed:
Modification:
added all method public in RecommendationRequestController;
create RejectionDto;
added method rejectRequest in RecommendationRequestService;

MikhailShap committed:
added the @Accessors(chain = true) annotation to all DTO

MikhailShap committed:
Modification:
changed the parameters of all methods for validation of the RecommendationRequestService class, instead of passing fields, passing an object;
create class RecommendationRequestServiceValidatorTest, create methods:
testRequesterNotExistsInDatabase,
testReceiverNotExistsInDatabase,
testRequestLimitIsLessThanSixMonths,
testSkillsNotExistenceInDatabase;

MikhailShap committed:
Modification:
added a check for an empty list before adding in create method in RecommendationRequestService;
added @Accessors(chain = true) in SkillRequest and User;
created RecommendationRequestServiceTest
created tests:
createWithValidData,
testGetRequestIfRequestNotExistsInDatabase,
testGetRequestValidData

MikhailShap committed:
Modification:
created CreatedAfterFilter;
created CreatedBeforeFilter;
added a check for the null list in method mapSkillRequestToSkillId for RecommendationRequestMapper;
changed the logic of using filters , made it more correct in method getRequests for RecommendationRequestService;
added method isFilterApplicable for Interface RequestFilter and for other implementations;

MikhailShap committed:
Modification:
CreatedAfterFilter and CreatedBeforeFilter made spring beans;

MikhailShap committed:
Modification for RecommendationRequestMapperTest:
removed the annotation @Spy from the object RecommendationRequestMapperImpl,I did the initialization manually;
made instead of the @BeforeAll @BeforeEach annotation for the setUp method , removed all static;
removed the @ExtendWith annotation;

MikhailShap committed:
Modification for RecommendationRequestServiceTest:
I made the logic of injecting a list with mocks;
created methods
testRejectRequestIfRecommendationRequestNotExistsInDatabase,testRejectRequestIfRecommendationRequestWithStatusIsNotPENDING,testRejectRequestValidData,testGetRequestsValidData

MikhailShap committed:
Modification:
created CreatedAfterFilterTest with methods:
isFilterApplicableFieldIsPresent;
isFilterApplicableFieldIsNotPresent;
passedFilter;
failedFilter;

MikhailShap committed:
Modification:
created CreatedBeforeFilterTest with methods:
isFilterApplicableFieldIsPresent;
isFilterApplicableFieldIsNotPresent;
passedFilter;
failedFilter;

MikhailShap committed:
Modification:
created CreatedBeforeFilterTest with methods:
isFilterApplicableFieldIsPresent;
isFilterApplicableFieldIsNotPresent;
passedFilter;
failedFilter;

MikhailShap committed:
Modification:
created StatusFilterTest with methods:
isFilterApplicableFieldIsPresent;
isFilterApplicableFieldIsNotPresent;
passedFilter;
failedFilter;

MikhailShap committed:
Modification:
moved the RecommendationRequestService to a separate recommendation folder

MikhailShap committed:
Modification:
added *.DS_Store in .gitignore

MikhailShap committed:
Удалил файлы .DS_Store и добавил их в .gitignore

MikhailShap committed:
Merge branch 'Medusa-master-stream7' into feature/BJS2-40449

MikhailShap committed:
Modification:
folder distribution

Shakhaev commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,199 @@
+        when(userRepository.findById(longThat(requesterAndReceiverIds)))
+                .thenReturn(Optional.of(new User()));
+
+        recommendationRequestService.create(recommendationRequestDto);
```
В данной строчке ты производишь создание запроса на рекомендацию, но, из-за того, что метод create в сервисе ничего не возвращает, поведение может быть непредсказуемым (а что, если ничего не создалось, или любая другая операция внутри метода create прошла неудачно). Возвращая в методе create саму DTO, с помощью которой ты создаешь запрос, в данном тесте ты можешь просто проиницилизировать объект: RecommendationRequestDto actualDto = recommendationRequestService.create(recommendationRequestDto); 
С какой целью это делать? А можно просто через assertEquals уже сравнить ожидаемую тобой DTO, которую ты отправляешь на создание и ту самую actualDto, которая тебе вернулась из метода create. Я сам таким образом делаю и выявляю в сервисе ошибочную логику (например, в своем сервисе я ожидал нормальную DTO, а на выходе из метода create получал объект null, что позволило мне исправить логику).

MikhailShap replied:
Нам же тут важно проверить , что вызывается метод save с нужным нам аргументом ,в данном случае это отправленная нами дто заммапленная в энтити , в принципе я это и проверяю в следующих строках,
        verify(recommendationRequestRepository).save(captor.capture());
        RecommendationRequest recommendationRequest = captor.getValue();
        assertEquals(recommendationRequestDto.getId(), recommendationRequest.getId());

MikhailShap replied:
(а что, если ничего не создалось, или любая другая операция внутри метода create прошла неудачно)
Если ты это про обычную работу, а не тесты , то мне кажется должна какая то ошибка выбрасываться , в этом моменте я не уверен , предполагаю.

Shakhaev replied:
Да, не заметил, что ты ниже проверяешь айдишники. Впринципе, для большей уверенности в совпадении, можно прям оба объекта дополнительно сравнить: assertEquals(recommendationRequestDto, recommendationRequest;

Shakhaev replied:
Да, действительно, ты уже сравниваешь ниже айдишники, не заметил. Что касаемо обычной работы метода, то тут тесты мне помогли выявить неправильную логику как раз за счет сравнения не отдельных полей, а целых объектов. Я ожидал получить именно ту DTO, которую я отправил в свой метод сервиса, а на выходе получил объект null. Не ошибку, а именно null. Долго разбирался, с чего вдруг там null появляется. По итогу, тест и сравнение ожидаемый/реальный результатов помогли мне выявить дыры и неровности в бизнес-логике. Поэтому, я стараюсь избегать void и возвращать DTO, чтобы тестировать, как минимум, было удобно.

Shakhaev commented on src/main/java/school/faang/user_service/service/recommendation/RecommendationRequestService.java:
```diff
@@ -0,0 +1,94 @@
+    private final List<RecommendationRequestFilter> recommendationRequestFilters;
+
+
+    public void create(RecommendationRequestDto recommendationRequestDto) {
```
Привет! Я предполагаю, что в данном методе тоже нужно вернуть DTO, поскольку в контроллере данный метод возвращает ответ клиенту  в виде RecommendationRequestDto. И еще будет удобно тестировать данный метод, чтобы возвращать ожидаемый объект для обеспечения большей прозрачности.

MikhailShap replied:
Я просто не знаю , на сколько это здесь нужно , у меня контроллер в принципе уже возвращает дто, или же нам надо после сохранения энтити получить объект, заммапить его в дто и уже его вернуть , это не уточнялось.
А по поводу тестов , ну мы же сами передаем объект и у нас изначально в тестах есть к нему доступ. ну и плюс вроде как главная цель метода , не обработать объект и вернуть его , а просто сохранить в бд.

Shakhaev replied:
Да, цель метода - сохранить сущность, полностью согласен. По сути, в тестах можно и каптором ловить нужный объект, классно, что ты его используешь. По моему мнению, можно иногда подзабыть заиспользовать каптор или запутаться в его применении, поэтому я лично особо с ним не парюсь и , как ты правильно написал, после сохранения энтити получить объект, заммапить его в дто и уже его вернуть.

Shakhaev commented on src/main/java/school/faang/user_service/filters/recommendation_request/StatusFilterRecommendation.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.Objects;
+
+@Component
+public class StatusFilterRecommendation implements RecommendationRequestFilter {
```
Думаю, стоит подумать над наименованием фильтров. Например, можно назвать как RecomendationStatusFilter.

MikhailShap replied:
Я переименовал вроде один класс интерфейс и мне автоматом IDEA переименовала реализации , я просто торопился и не стал исправлять , так оставил

Shakhaev commented on src/test/java/school/faang/user_service/mapper/RecommendationRequestMapperTest.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Test
+    void toDTO() {
```
В тестах лучше всегда использовать глагол test в начале названия тестового метода.

MikhailShap replied:
Согласен , что то пропустил

Shakhaev commented on src/test/java/school/faang/user_service/service/recommendation/RecommendationRequestServiceTest.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Test
+    void createWithValidData() {
```
Здесь тоже лучше приписать глагол test в начале.

flornana commented on src/main/java/school/faang/user_service/dto/filter/RequestFilterDto.java:
```diff
@@ -0,0 +1,33 @@
+    private Long requesterId;
+    private Long receiverId;
+    @NotNull
+    private RequestStatus status;
```
Тут тоже можно message добавить

MikhailShap replied:
Чуть переписал логику , сделал ее более корректной , поэтому эта аннотация там ни к чему , убрал ее

MikhailShap committed:
Modification:
add TODO:

MiraGlaceon commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,34 @@
+
+@Controller
+@RequiredArgsConstructor
+public class RecommendationRequestController {
```
Нужно переписать на RestController

MikhailShap committed:
Modification:
refactoring name for filter classes;
added the prefix "test" to the filter test methods;
added for RecommendationRequestController:
Annotations for class:
@Slf4j,
@RequestMapping("/recommendation");
Annotations for mapping queries and the
necessary annotations for parameters , for all methods;
Logging for all methods;

MikhailShap committed:
Modification:
removed ToDO List

MikhailShap committed:
Modification:
added logging for classes:
RecommendationCreatedAfterFilter,
RecommendationCreatedBeforeFilter,
RecommendationRequestService,
RecommendationRequestServiceValidator,
RecommendationStatusFilter;

MikhailShap committed:
Modification:
removed the @NOTNULL annotation from the class RequestFilterDto

MikhailShap committed:
Modification:
removed the @NOTNULL annotation from the class RequestFilterDto, installed an error check EntityNotFoundExceptionWithID.class

MikhailShap committed:
Modification for RecommendationRequestController:
created tests for all methods;
removed @Valid from method getRecommendationRequests;
Modification for RecommendationRequestDto:
created exception message for skillIds;

MikhailShap committed:
Merge branch 'Medusa-master-stream7' into feature/BJS2-40449

MikhailShap committed:
Modification:
added validation , id cannot be a negative number
